src/github.com/getlantern/flashlight/http/http.go

```diff
@@ -0,0 +1,167 @@
+package http
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/getlantern/flashlight/client"
+	"github.com/getlantern/flashlight/config"
+	"github.com/getlantern/flashlight/util"
+	"github.com/getlantern/golog"
+	"github.com/getlantern/whitelist"
+	"github.com/skratchdot/open-golang/open"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/getlantern/tarfs"
+)
+
+const (
+	UiUrl = "http://%s"
+	UiDir = "src/github.com/getlantern/ui/app"
+)
+
+var (
+	log = golog.LoggerFor("http")
+)
+
+type JsonResponse struct {
+	Error     string   `json:"Error, omitempty"`
+	Whitelist []string `json:"Whitelist, omitempty"`
+	Original  []string `json:"Original, omitempty"`
+}
+
+type WhitelistHandler struct {
+	http.HandlerFunc
+	whitelist *whitelist.Whitelist
+	wlChan    chan *whitelist.Config
+}
+
+func sendJsonResponse(w http.ResponseWriter, response *JsonResponse, indent bool) {
+
+	enc := json.NewEncoder(w)
+	err := enc.Encode(response)
+	if err != nil {
+		log.Errorf("error sending json response %v", err)
+	}
+}
+
+func setResponseHeaders(w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token")
+	w.Header().Set("Access-Control-Allow-Credentials", "True")
+	w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
+
+}
+
+func (wlh WhitelistHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var response JsonResponse
+
+	setResponseHeaders(w)
+
+	switch r.Method {
+	case "OPTIONS":
+		// return if it's a preflight OPTIONS request
+		// this is mainly for testing the UI when it's running on
+		// a separate port
+		return
+	case "POST":
+		// update whitelist
+		decoder := json.NewDecoder(r.Body)
+		var entries []string
+		err := decoder.Decode(&entries)
+		if err != nil {
+			log.Error(err)
+			response.Error = fmt.Sprintf("Error decoding whitelist: %q", err)
+		} else {
+			wl := wlh.whitelist.UpdateEntries(entries)
+			copy := wlh.whitelist.Copy()
+			log.Debug("Propagating whitelist changes..")
+			wlh.wlChan <- copy
+			response.Whitelist = wl
+		}
+	case "GET":
+		response.Whitelist = wlh.whitelist.GetEntries()
+	default:
+		log.Debugf("Received %s", response.Error)
+		response.Error = "Invalid whitelist HTTP request"
+		response.Whitelist = nil
+	}
+	sendJsonResponse(w, &response, false)
+}
```

Comment:
If I understand this protocol correctly, the client posts the entire list every time and the back-end calculates the deltas.  This protocol will have problems if the user uses multiple windows at once.  After one window updates, the other window will have a stale proxiedsites list, and if the user makes an update there, it will overwrite whatever the user did in the other window.  However, since the other window doesn't update, there won't be anything to make that clear to the user.

A safer protocol is to have the UI calculate the deltas and post those.  The deltas are really what the user is intending.  If the user submits conflicting deltas from two different windows, that's okay because they both still represent the user's intention at the time.  A similar problem arises if someone edits the yaml file manually.

I think the best protocol would be something like this:

- Use a websocket
- Back-end posts whitelist deltas to the websocket.  On initial connection, the entire whitelist is posted as a series of adds.  For performance, we should be able to batch multiple sites in a single delta (e.g. add sites 1, 2 and 3).
- When a user makes changes to the whitelist, post those changes to the websocket as deltas (again, support batching for performance)
- Whenever the whitelist has changed, post the deltas to all connected UIs (note, this should be driven by the YAML config having been changed, not by the HTTP handler)

This will allow multiple windows and the config on disk to remain in sync.  It will also be a more useful protocol once we have an extension that allows people to add/remove the current site from the whitelist, as they can use the same API and just have a batch of 1.